src/OrchardCore.Modules/OrchardCore.Lists/GraphQL/ListQueryObjectType.cs

```diff
@@ -22,36 +26,36 @@ public ListQueryObjectType(IStringLocalizer<ListQueryObjectType> T)
                 .Description("the content items")
-                .Argument<IntGraphType, int>("first", "the first n elements (10 by default)", 0)
+                .Argument<IntGraphType, int>("first", "the first n elements (10 by default)", 10)
                 .Argument<IntGraphType, int>("skip", "the number of elements to skip", 0)
                 .ResolveAsync(async g =>
                 {
-                    var context = (GraphQLContext)g.UserContext;
-                    var session = context.ServiceProvider.GetService<ISession>();
+                    var serviceProvider = g.ResolveServiceProvider();
+                    var session = serviceProvider.GetService<ISession>();
+                    var accessor = serviceProvider.GetRequiredService<IDataLoaderContextAccessor>();
 
-                    var query = session.Query<ContentItem>()
-                        .With<ContainedPartIndex>(x => x.ListContentItemId == g.Source.ContentItem.ContentItemId)
-                        .With<ContentItemIndex>(x => x.Published)
-                        .OrderByDescending(x => x.CreatedUtc);
+                    var dataLoader = accessor.Context.GetOrAddCollectionBatchLoader<string, ContentItem>("ContainedPublishedContentItems", x => LoadPublishedContentItemsForListAsync(x, session));
 
-                    // Apply a default limit
-                    var pagedQuery = query.Take(10);
-
-                    var skip = g.GetArgument<int>("skip");
-                    var first = g.GetArgument<int>("first");
-
-                    if (skip > 0)
-                    {
-                        pagedQuery = pagedQuery.Skip(skip);
-                    }
-
-                    if (first > 0)
-                    {
-                        pagedQuery = pagedQuery.Take(first);
-                    }
-
-                    return await query.ListAsync();
+                    return (await dataLoader.LoadAsync(g.Source.ContentItem.ContentItemId))
+                                .Skip(g.GetArgument<int>("skip"))
+                                .Take(g.GetArgument<int>("first"));
                 });
         }
+
+        private static async Task<ILookup<string, ContentItem>> LoadPublishedContentItemsForListAsync(IEnumerable<string> contentItemIds, ISession session)
+        {
+            if (contentItemIds is null || !contentItemIds.Any())
+            {
+                 return new Dictionary<string, ContentItem>().ToLookup(k => k.Key, v => v.Value);
+            }
+
+            var query = await session.Query<ContentItem>()
+                                     .With<ContentItemIndex>(ci => ci.Published)
+                                     .With<ContainedPartIndex>(cp => cp.ListContentItemId.IsIn(contentItemIds))
+                                     .OrderBy(o => o.Order)
+                                     .ListAsync();
+
+            return query.ToLookup(k => k.As<ContainedPart>().ListContentItemId);
+        }
     }
-}
+}
```

Comment:
nit

Reply:
dammit! dont mind doing it but can i ask the reasoning ;p